src/Abstractions/Entities/TaskOrchestrationEntityFeature.cs

```diff
@@ -0,0 +1,97 @@
+    /// </summary>
+    /// <param name="id">The target entity.</param>
+    /// <param name="operationName">The name of the operation.</param>
+    /// <param name="input">The operation input.</param>
+    /// <param name="options">The call options.</param>
+    /// <returns>
+    /// A task that completes when the operation has been completed. Or in the signal-only case, a task that is either
+    /// already complete or completes when the operation has been enqueued.
+    /// </returns>
+    public abstract Task CallEntityAsync(
+        EntityInstanceId id, string operationName, object? input = null, CallEntityOptions? options = null);
+
+    /// <summary>
+    /// Calls an operation on an entity and waits for it to complete. Does not wait for completion if
+    /// <see cref="CallEntityOptions.Signal"/> is set on <paramref name="options"/>.
+    /// </summary>
+    /// <param name="id">The target entity.</param>
+    /// <param name="operationName">The name of the operation.</param>
+    /// <param name="options">The call options.</param>
+    /// <returns>
+    /// A task that completes when the operation has been completed. Or in the signal-only case, a task that is either
+    /// already complete or completes when the operation has been enqueued.
+    /// </returns>
+    public virtual Task CallEntityAsync(EntityInstanceId id, string operationName, CallEntityOptions? options)
+        => this.CallEntityAsync(id, operationName, null, options);
+
+    /// <summary>
+    /// Acquires one or more entity locks.
+    /// </summary>
+    /// <param name="entityIds">The entity IDs to lock.</param>
+    /// <returns>An async-disposable which can be disposed to release the lock.</returns>
+    public abstract Task<IAsyncDisposable> LockEntitiesAsync(IEnumerable<EntityInstanceId> entityIds);
+
+    /// <summary>
+    /// Acquires one or more entity locks.
+    /// </summary>
+    /// <param name="entityIds">The entity IDs to lock.</param>
+    /// <returns>An async-disposable which can be disposed to release the lock.</returns>
+    public virtual Task<IAsyncDisposable> LockEntitiesAsync(params EntityInstanceId[] entityIds)
+        => this.LockEntitiesAsync((IEnumerable<EntityInstanceId>)entityIds); // let the implementation decide how to handle nulls.
+
+    /// <summary>
+    /// Gets a value indicating whether any entity locks are owned by this instance.
+    /// </summary>
+    /// <param name="entityIds">The list of locked entities.</param>
+    /// <returns>True if any locks are owned, false otherwise.</returns>
+    public abstract bool HasEntityLocks(out IReadOnlyList<EntityInstanceId> entityIds);
+
+    /// <summary>
+    /// Gets a value indicating whether any entity locks are owned by this instance.
+    /// </summary>
+    /// <returns>True if any locks are owned, false otherwise.</returns>
+    public virtual bool HasEntityLocks() => this.HasEntityLocks(out _);
```

Comment:
this is not entirely right in terms of functionality.

These are the pieces of information we want to expose:

-  whether the orchestrator is currently inside a critical section
-  if inside a critical section, which entities are currently available for making calls

where a lock is available to make the call if (a) it was locked at the beginning of the critical section, and (b) there is not currently a call to that entity already pending. 

So, whether or not we are in a critical section cannot be determined simply by checking if the list is empty - it can be empty even if we are inside a critical section.

Reply:
This is what I had before:

```
 /// <summary>
 /// Whether this orchestration is currently inside a critical section.
 /// </summary>
 public bool IsInsideCriticalSection => this.criticalSectionId != null;   

 /// <summary>
 /// Enumerate all the entities that are available for calling from within a critical section. 
 /// This set contains all the entities that were locked prior to entering the critical section,
 /// and for which there is not currently an operation call pending.
 /// </summary>
 /// <returns>An enumeration of all the currently available entities.</returns>
 public IEnumerable<(string entityName, string entityKey)> GetAvailableEntities()
 {
     if (this.IsInsideCriticalSection)
     {
         foreach(var e in this.availableLocks)
         {
             yield return (e.Name, e.Key);
         }
     }
 }
```

it is not perfect but it should give you an idea of how it is intended to work.
Another option would be to create 2 overloads of `IsInsideCriticalSection`, where one of them takes an out enumerable parameter. That is probably a little better than the above `GetAvailableEntities`.

Reply:
Would it be sufficient to rename `HasEntityLocks` to `InCriticalSection`?